Import swift_attr(“@Sendable”) for result types

beccadax opened the pull request:
If `__attribute__((swift_attr(“@Sendable”)))` is applied to an ObjC method, ObjC property, C field, C variable, or C function, we will now make its result type `Sendable`.

For some entities, this technically had a different behavior previously, because `@Sendable` can be applied to `func`s to indicate that their interface type should be `@Sendable`. We don’t really want people to use this functionality on non-local functions, so we can safely remove it.

This isn’t quite interacting with `Unmanaged` the way we’d want, but I’ve included test cases for the current behavior with FIXME comments.

Fixes rdar://90491024.

As part of this work, I have also made several NFC changes:

* Threaded an `llvm::function_ref<void(Diagnostic&&)>` parameter through `importType()` and its implementation, so we can emit import diagnostics on whatever declaration we happen to be converting a given type to import.

* Refactored `applyParamAttributes()` into `ImportTypeAttrs`, a more general mechanism for collecting certain attributes from the declaration being imported and passing them to `importType()`.

* Eliminated the special case for `swift_attr("@Sendable")` on a typedef—it has been subsumed by the general `ImportTypeAttrs` mechanism.

beccadax committed:
[NFC] Factor out ImportTypeAttrs

Preparation for integrating the logic remaining in `applyParamAttributes()` into `importType()`.

beccadax committed:
[NFC] Eliminate `applyParamAttributes()`

beccadax committed:
Import swift_attr(“@Sendable”) for result types

If `__attribute__((swift_attr(“@Sendable”)))` is applied to an ObjC method, ObjC property, C field, C variable, or C function, we will now make its result type `Sendable`.

For some entities, this technically had a different behavior previously, because `@Sendable` can be applied to `func`s to indicate that their interface type should be `@Sendable`. We don’t really want people to use this functionality on non-local functions, so we can safely remove it.

This isn’t quite interacting with `Unmanaged` the way we’d want, but I’ve included test cases for the current behavior with FIXME comments.

Fixes rdar://90491024.

beccadax commented:
Tagging @NuriAmari specifically to review commit 00724f5.

beccadax commented:
@swift-ci please smoke test

beccadax commented:
@swift-ci please test source compatibility

NuriAmari commented:
I've only looked at https://github.com/apple/swift/commit/00724f50290998ee26824a352d8bee91805f09dd, but it looks great to me! Thanks for keeping me in the loop :smile:

beccadax commented:
Looks like the source compat suite was busted, but it may be fixed now.

@swift-ci please test source compatibility

beccadax committed:
[NFC] Thread import diagnostics through ImportType

`ClangImporter::Implementation::importType()` and associated parts of the importer are now passed an `llvm::function_ref` they can use to emit import diagnostics on the declaration they’re importing, and the `ImportDiagnosticAdder` helper class provides a convenient way to construct such a function.

This capability is not actually *used* in this commit—we are simply threading this function through the importer—so there is no change in behavior.

beccadax committed:
[NFC] Integrate ImportTypeAttrs into importType()

beccadax commented:
@swift-ci smoke test

beccadax commented:
Rebased to resolve a conflict with #41782.

DougGregor commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -1866,6 +1885,9 @@ class GetSendableType :
     // want to make their generic parameters `Sendable`.
+    // FIXME: `Unmanaged` ought to be here too, but `AnyObject & Sendable`
+    //        doesn't satisfy the generic parameter's `AnyObject` requirement.
+    //        (rdar://90946615)
```
Not a big deal, but I prefer not to have Radar numbers in source code, because they're not useful outside of Apple. The FIXME text already suffices to express the issue.

DougGregor commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -1985,22 +2004,42 @@ Type ClangImporter::Implementation::applyParamAttributes(
+      });
+    } else {
+      // If we can't use @MainActor, fall back to at least using @Sendable.
+      attrs |= ImportTypeAttr::Sendable;
```
I find this to be a strange fallback. `MainActor` is going to be there unless some strange module-loading error occurred and we're recovering from that. I don't think adding `@Sendable` is necessary along this path.

DougGregor commented on lib/ClangImporter/ImporterImpl.h:
```diff
@@ -1681,6 +1736,22 @@ class LLVM_LIBRARY_VISIBILITY ClangImporter::Implementation
   }
 };
 
+class ImportDiagnosticAdder {
```
This is a nice structural cleanup, thanks!

DougGregor reviewed (approved):
This looks great! A few minor comments, but it's all small post-merge cleanup.
